In [1]:
import json
import math
import os
import sys

from geographiclib.geodesic import Geodesic

geod = Geodesic.WGS84  # define the WGS84 ellipsoid

# Make sure local modules can be imported
module_path_root = os.path.abspath(os.pardir)
if module_path_root not in sys.path:
    sys.path.append(module_path_root)

In [2]:
# Specify the locality whose OSM file we want to load

# OSM Bike Lanes only
#locality = 'Locality_OSM_cycleway_Mount_Eliza.geojson'

# OSM all routes
locality = 'Locality_OSM_all_Mount_Eliza.geojson'

In [3]:
geojson_filename = os.path.join(os.pardir, 'data_sources', locality)

with open(geojson_filename) as geojson_file:
    geojson_data = json.load(geojson_file)
    geojson_file.close()

In [4]:
def expand_points_pair(lat1, lon1, lat2, lon2, interval_metres=10):
    points_and_bearings = []
    
    bearing = geod.Inverse(lat1, lon1, lat2, lon2)['azi1']
    
    line = geod.InverseLine(lat1, lon1, lat2, lon2)
    num_steps = int(math.ceil(line.s13 / interval_metres))
    for step_i in range(num_steps + 1):
        s = min(interval_metres * step_i, line.s13)
        g = line.Position(s, Geodesic.STANDARD | Geodesic.LONG_UNROLL)
        points_and_bearings.append([g['lat2'], g['lon2'], bearing])
    
    return points_and_bearings

In [5]:
def expand_points_feature(feature, interval_metres=10):
    points_and_bearings = []
    
    geometry    = feature['geometry']
    coordinates = geometry['coordinates']
    
    #print('Coordinate Count: ' + str(len(coordinates)))
    
    for point_i in range(len(coordinates)-1):
        point1 = coordinates[point_i]
        point2 = coordinates[point_i+1]
        
        # Get points at regular intervals between the pair of points
        span = expand_points_pair(float(point1[1]), float(point1[0]), float(point2[1]), float(point2[0]), interval_metres)
        
        # If this is not the last pair of points, remove the LAST point
        # (so that we don't repeat it unnecessarily
        if point_i < len(coordinates)-2:
            span.pop()
            
        # Add these points to the overall line for the "feature" or "way"
        points_and_bearings = points_and_bearings + span
    
    return points_and_bearings

In [6]:
features = geojson_data['features']

locations = []

for idx, feature in enumerate(features):
    try:
        name = feature['properties']['name']
    except Exception:
        name = '?'
        
    print('%05d %s' % (idx, name))

    locations.append(expand_points_feature(feature))

00000 ?
00001 Wooralla Drive
00002 Humphries Road
00003 Canadian Bay Road
00004 Walkers Road
00005 Baden Powell Drive
00006 Mornington Golf Club
00007 Shotton Road
00008 Oakbank Road
00009 Canadian Bay Road
00010 Mount Eliza Way
00011 ?
00012 Mount Eliza Way
00013 Mount Eliza Way
00014 Mount Eliza Way
00015 ?
00016 Sunnyside Road
00017 ?
00018 Wooralla Drive
00019 ?
00020 Mountain View Road
00021 Mountain View Road
00022 ?
00023 Mountain View Road
00024 Quinns Parade
00025 Eliza Drive
00026 Quinns Parade
00027 Tower Road
00028 Plover Court
00029 Tern Grove
00030 Yewers Avenue
00031 Dalry Court
00032 Cormorant Place
00033 Stewart Street
00034 Petrel Close
00035 Sandpiper Way
00036 Cameron Way
00037 Duncan Close
00038 Kolora Crescent
00039 Kunyung Road
00040 Acheron Avenue
00041 Rutland Avenue
00042 Wimborne Avenue
00043 Rugby Court
00044 Wimbledon Avenue
00045 Rannoch Avenue
00046 Roehampton Crescent
00047 Camborne Avenue
00048 Wimborne Avenue
00049 Roborough Avenue
00050 Redbourne Aven

00538 ?
00539 ?
00540 ?
00541 ?
00542 ?
00543 ?
00544 ?
00545 ?
00546 ?
00547 ?
00548 ?
00549 ?
00550 ?
00551 ?
00552 Ranelagh Beach
00553 ?
00554 ?
00555 ?
00556 ?
00557 ?
00558 ?
00559 Lorikeet Reserve
00560 Manna Hill Court
00561 Saint Kilda Street
00562 Wooralla Drive
00563 Parklands Grove
00564 Parklands Grove
00565 Bird Street
00566 Eumeralla Grove
00567 Orrong Grove
00568 Victory Place
00569 Heath Road
00570 Boxmoor Court
00571 ?
00572 Wimbledon Avenue
00573 Moorgate Avenue
00574 Granya Grove
00575 Narrung Road
00576 Village Lane
00577 Keraboite Court
00578 Tintagel Court
00579 Cypress Point Drive
00580 Denistoun Avenue
00581 Orchard Lane
00582 Charles Street
00583 Drumdoune Drive
00584 Jerula Avenue
00585 Bonnyview Road
00586 Eliza Terrace
00587 ?
00588 Kirkstall Close
00589 Humphries Road
00590 FR Mellor & Son siding
00591 ?
00592 ?
00593 ?
00594 ?
00595 ?
00596 ?
00597 ?
00598 McAuliffe Close
00599 Manyung Court
00600 Moondah Drive
00601 ?
00602 Moorooduc Highway
00603 ?
0060

01305 ?
01306 ?
01307 ?
01308 ?
01309 ?
01310 ?
01311 ?
01312 ?
01313 ?
01314 ?
01315 ?
01316 ?
01317 ?
01318 ?
01319 ?
01320 ?
01321 ?
01322 ?
01323 ?
01324 ?
01325 ?
01326 ?
01327 ?
01328 ?
01329 ?
01330 ?
01331 ?
01332 ?
01333 ?
01334 ?
01335 ?
01336 ?
01337 ?
01338 Nepean Highway
01339 ?
01340 ?
01341 ?
01342 ?
01343 ?
01344 ?
01345 ?
01346 Jessie White Creek aka Earimil Creek
01347 Jessie White Creek
01348 ?
01349 Jessie White Creek aka Earimil Creek
01350 ?
01351 Jessie White Creek aka Earimil Creek
01352 ?
01353 ?
01354 ?
01355 ?
01356 Earimil Creek
01357 ?
01358 ?
01359 ?
01360 ?
01361 ?
01362 The Ridge Reserve
01363 ?
01364 Quinn's Parade Reserve
01365 ?
01366 ?
01367 Leindan Court Reserve
01368 ?
01369 ?
01370 ?
01371 ?
01372 ?
01373 Sunset Crescent Reserve
01374 Allambi Court Reserve
01375 ?
01376 ?
01377 ?
01378 ?
01379 ?
01380 ?
01381 ?
01382 ?
01383 ?
01384 ?
01385 ?
01386 ?
01387 ?
01388 ?
01389 ?
01390 ?
01391 ?
01392 John H Butler Reserve
01393 ?
01394 ?
01395 ?
01396 

In [7]:
print('Way Count:   ' + str(len(locations)))
print('Point Count: ' + str(sum([ len(location) for location in locations])))

Way Count:   1804
Point Count: 75740


In [8]:
#print(locations[0])

1104 points if checking only OSM cycleway routes for Mount Eliza
75740 points if checking all routes

$7 USD per 1000 images
4 images per point

==> $303 USD for the one town